tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc

```diff
@@ -1886,6 +1886,105 @@ bool StridedSliceOp::GetSlicedBoundRanges(
+  //   %height = tf.StridedSlice(%shape, 1, 2, 1)
+  //
+  // In this case %height can be replaced with a constant 1.
+  //
+  // Or,
+  //
+  //   %shape  = tf.Shape(%arg)                   // %arg: tensor<?x2x3x1xf32>
+  //   %spatial_shape = tf.StridedSlice(%shape, 1, 3, 1)
+  //
+  // In this case %spatial_shape can be replaced with a constant [2, 3].
+
+  // Input to strided slice op is defined by shape operation.
+  auto shape_op = input().getDefiningOp<ShapeOp>();
+  if (!shape_op) {
+    return {};
+  }
+
+  // `begin`, `end` and `strides` should be constant in order to infer static
+  // dimension.
+  DenseIntElementsAttr begin_attr, end_attr, strides_attr;
+  if (!matchPattern(begin(), m_Constant(&begin_attr)) ||
+      !matchPattern(end(), m_Constant(&end_attr)) ||
+      !matchPattern(strides(), m_Constant(&strides_attr)) ||
+      begin_attr.getNumElements() != 1 || end_attr.getNumElements() != 1 ||
+      strides_attr.getNumElements() != 1) {
+    return {};
+  }
+
+  auto tensor_ty = shape_op.input().getType().dyn_cast<RankedTensorType>();
+  auto output_ty = output().getType().dyn_cast<RankedTensorType>();
+  // Only ranked tensor can be folded.
+  if (!tensor_ty || !output_ty) {
+    return {};
+  }
+
+  int64_t rank = tensor_ty.getRank();
+  int64_t begin_int = begin_attr.getValue<APInt>(0).getSExtValue();
+  int64_t end_int = end_attr.getValue<APInt>(0).getSExtValue();
+  int64_t strides_int = strides_attr.getValue<APInt>(0).getSExtValue();
+
+  // Canonicalize `begin` and `end` in case of negative index.
+  if (begin_int < 0) begin_int += rank;
+  if (end_int < 0) end_int += rank;
+
+  // Create `begin` and `end` from `*_mask`. Note that we don't care about
+  // `new_axis_mask` and `shrink_axis_mask` as they can be inferred from
+  // `output_ty`.
+  if (begin_mask() == 1) {
+    begin_int = (strides_int > 0) ? 0 : rank - 1;
+  }
+  if (end_mask() == 1) {
+    end_int = (strides_int > 0) ? rank : -1;
+  }
+  if (ellipsis_mask() == 1) {
+    begin_int = 0;
+    end_int = rank;
+  }
+
+  // It's possible that begin and end are out of bound. See
+  // https://docs.python.org/3/library/stdtypes.html#common-sequence-operations.
+  if (strides_int > 0) {
+    begin_int = std::min(begin_int, rank);
+    end_int = std::min(end_int, rank);
+  } else {
+    begin_int = std::min(begin_int, rank - 1);
+    end_int = std::min(end_int, rank - 1);
+  }
+
+  SmallVector<int64_t, 2> sub_shape;
+  // Only handle cases that have something to slice to avoid infinite for-loop.
+  if ((end_int > begin_int && strides_int > 0) ||
+      (end_int < begin_int && strides_int < 0)) {
+    // Extract sub-shape only if all of those dimensions are static.
+    for (int64_t i = begin_int; (strides_int > 0) ? i < end_int : i > end_int;
+         i += strides_int) {
+      if (tensor_ty.isDynamicDim(i)) {
+        return {};
+      }
+      sub_shape.push_back(tensor_ty.getDimSize(i));
+    }
+  }
+
+  // Down-cast to 32 bit int if needed.
```

Comment:
I may suggest to write it in a templated way:

```
  auto buildDenseElementAttr = [&]<class T>() -> OpFoldResult {
    SmallVector<T> sub_shape;
    // Only handle cases that have something to slice to avoid infinite for-loop.
    if ((end_int > begin_int && strides_int > 0) ||
         (end_int < begin_int && strides_int < 0)) {
      // Extract sub-shape only if all of those dimensions are static.
      for (int64_t i = begin_int; (strides_int > 0) ? i < end_int : i > end_int;
            i += strides_int) {
        if (tensor_ty.isDynamicDim(i))
          return {};
        sub_shape.push_back(tensor_ty.getDimSize(i));
      }
    }
    return DenseIntElementsAttr::get(output_ty, sub_shape);
  };
  if (output_ty.getElementType().isInteger(32))
    return buildDenseElementAttr<int32_t>();
  return buildDenseElementAttr<int64_t>();
```

Reply:
Cool! Forget c++14 allow this.

Reply:
Seems that it doesn't work in MSVC. Let me rollback it :(

Reply:
Of course it doesn't... :(